litebird_sim/mapmaking/destriper.py

```diff
@@ -0,0 +1,2102 @@
+        #
+        #       |  1/a             0      0  |
+        # L⁻¹ = | −b/ac           1/c     0  |
+        #       | (be − cd)/acf  −e/cf   1/f |
+        #
+        # We assign its coefficients to the following letters:
+        #
+        #       | g 0 0 |
+        # L⁻¹ = | h i 0 |
+        #       | j k m |
+        g = 1 / a
+        h = -b / (a * c)
+        i = 1 / c
+        j = (b * e - c * d) / (a * c * f)
+        k = -e / (c * f)
+        m = 1 / f
+
+        # Now it's time for us to compute M⁻¹. We can do this quickly:
+        #
+        # M⁻¹ = (L L^T)⁻¹ = (L^T)⁻¹ L⁻¹ = (L⁻¹)^T L⁻¹
+        #
+        # where we used the fact that (AB)⁻¹ = B⁻¹ A⁻¹, and that
+        # (L^T)⁻¹ = (L⁻¹)^T. The product of (L⁻¹)^T and L⁻¹ is
+        #
+        # | g h i |   | g 0 0 |   | g²+h²+j²  hi+jk  jm |
+        # | 0 i k | · | h i 0 | = |  hi+jk   i²+k²   km |
+        # | j k m |   | j k m |   |   jm      km     m² |
+        result[ipix, 0, 0] = g**2 + h**2 + j**2
+        result[ipix, 1, 0] = h * i + j * k
+        result[ipix, 0, 1] = result[ipix, 1, 0]
+        result[ipix, 1, 1] = i**2 + k**2
+        result[ipix, 2, 0] = j * m
+        result[ipix, 0, 2] = result[ipix, 2, 0]
+        result[ipix, 2, 1] = k * m
+        result[ipix, 1, 2] = result[ipix, 2, 1]
+        result[ipix, 2, 2] = m**2
+
+
+@dataclass
+class NobsMatrix:
+    """
+    A class containing the N_obs matrix described in Kurki-Suonio et al. (2009)
+
+    The matrix is used to “solve” for the value of the Stokes parameters I/Q/U
+    per each pixel given a set of intensity measurements for that pixel.
+
+    The N_obs matrix is a block-diagonal matrix where each 3×3 block corresponds
+    to one pixel. (See Eq. 10 in Kurki-Suonio et al.) However, since the matrix
+    is symmetric, this class only stores its lower-triangular part, and the
+    ``nobs_matrix`` field is thus a 2D array of shape ``(6, N_pix)``, with
+    ``N_pix`` the number of pixels in the map.
+
+    The array `valid_pixel` is an array of ``N_pix`` booleans, telling whether
+    the corresponding pixel was observed enough times to make the problem of
+    reconstructing the Stokes I/Q/U components solvable or not.
+
+    If `is_cholesky` is true, the `nobs_matrix` contains the coefficients of
+    the lower triangular L such that L·L^t = M, where M is the 3×3 submatrix
+    in Eq. (10). This is extremely efficient, as in this way it is trivial
+    to invert matrix M to solve for the three Stokes parameter. And since L
+    is lower-triangular, 6 elements are still enough to keep it.
+    """
+
+    # Shape: (Npix, 6)
+    nobs_matrix: npt.NDArray
+    # Shape: (Npix,) (array of Boolean flags)
+    valid_pixel: npt.NDArray
+    # True if `nobs_matrix` contains the Cholesky decomposition of M_i
+    is_cholesky: bool
+
+    @property
+    def nbytes(self) -> int:
+        """Return the number of bytes used by this object"""
+        return self.nobs_matrix.nbytes + self.valid_pixel.nbytes
```

Comment:
Yes, I willingly left it out, as it is a Python object and it's thus unclear how to get the actual number of bytes used. (No, it's not just a byte, because it is wrapped around a C `PyObject` struct.)